File: src/lib/payment.ts
export const PAYMENT_METHODS = [
  { id: "pay_on_delivery", label: "Pay on delivery" },
  { id: "card", label: "Card" },
  { id: "apple_pay", label: "Apple Pay" },
  { id: "google_pay", label: "Google Pay" },
] as const;

export type PaymentMethodId = (typeof PAYMENT_METHODS)[number]["id"];
export type CardBrand = "Visa" | "Mastercard" | "Amex" | "Card";
const CARD_BRANDS: readonly CardBrand[] = ["Visa", "Mastercard", "Amex", "Card"];

export function isPaymentMethodId(value: unknown): value is PaymentMethodId {
  return PAYMENT_METHODS.some((method) => method.id === value);
}

export function parsePaymentMethod(value: unknown): PaymentMethodId | null {
  return isPaymentMethodId(value) ? value : null;
}

export function cardDigits(value: string): string {
  return value.replace(/\D/g, "");
}

export function detectCardBrand(value: string): CardBrand {
  const digits = cardDigits(value);
  if (digits.startsWith("4")) return "Visa";
  if (/^5[1-5]/.test(digits)) return "Mastercard";
  if (/^3[47]/.test(digits)) return "Amex";
  return "Card";
}

export function cardLast4(value: string): string | null {
  const digits = cardDigits(value);
  return digits.length >= 4 ? digits.slice(-4) : null;
}

export function isCardBrand(value: unknown): value is CardBrand {
  return typeof value === "string" && CARD_BRANDS.some((brand) => brand === value);
}

export function formatCardPaymentLabel({
  brand,
  last4,
}: {
  brand: CardBrand;
  last4: string;
}): string | null {
  if (!/^\d{4}$/.test(last4)) return null;
  return `Card · ${brand} ending ${last4}`;
}

export function paymentMethodLabel(method: PaymentMethodId): string {
  return PAYMENT_METHODS.find((option) => option.id === method)?.label ?? "Payment";
}

export function paymentStatusLabel(paymentMethod: string): "Paid (demo)" | "Due on delivery" {
  return paymentMethod === "Pay on delivery" ? "Due on delivery" : "Paid (demo)";
}

File: backend/app/domain/payment.py
"""Payment method labels (ported from src/lib/payment.ts)."""

from __future__ import annotations

from typing import Literal, Optional

PAYMENT_METHODS = (
    ("pay_on_delivery", "Pay on delivery"),
    ("card", "Card"),
    ("apple_pay", "Apple Pay"),
    ("google_pay", "Google Pay"),
)

PaymentMethodId = Literal["pay_on_delivery", "card", "apple_pay", "google_pay"]
CardBrand = Literal["Visa", "Mastercard", "Amex", "Card"]

_CARD_BRANDS = frozenset({"Visa", "Mastercard", "Amex", "Card"})
_METHOD_IDS = frozenset(method_id for method_id, _ in PAYMENT_METHODS)
_METHOD_LABELS = dict(PAYMENT_METHODS)


def parse_payment_method(value: object) -> Optional[PaymentMethodId]:
    if isinstance(value, str) and value in _METHOD_IDS:
        return value  # type: ignore[return-value]
    return None


def is_card_brand(value: object) -> bool:
    return isinstance(value, str) and value in _CARD_BRANDS


def format_card_payment_label(*, brand: str, last4: str) -> Optional[str]:
    if not isinstance(last4, str) or len(last4) != 4 or not last4.isdigit():
        return None
    if not is_card_brand(brand):
        return None
    return f"Card · {brand} ending {last4}"


def payment_method_label(method: PaymentMethodId) -> str:
    return _METHOD_LABELS.get(method, "Payment")
